### DISCUSSION TITLE
Wrong <EOS> output in inverse sequence design

### DISCUSSION THREAD

**Discussion by sirius777coder** [facebookresearch/esm#282] [Q&A] [answered]
Great work of ESM-IF1 and ESM2. When I use ESM-IF1 to design some sequences, i may get sequences with <eos> token in the end of this sequence. Is this a bug of the code? Here is an example.

**Comment by sirius777coder** [comment]
I may get a \<EOS> token in the end of the sequence

**Comment by naailkhan28** [comment]
I've seen this in my experiments with ESM-IF too, usually using a temperature value of 2 or higher. I can only imagine this is because a higher temperature makes the probability distribution more uniform, increasing the chances of incorporating a "rare" non amino acid token. 

I've experimented with implementing alternative sampling methods for ESM-IF and adding a check to catch these incorrect tokens would be easy - is this something we should think about adding? Happy to give it a go.

**Reply by sirius777coder** [reply]
Great answer. I think it may be a reason. But I use the temperature=1e-6 to design all of my sequences, and I still find these tokens. Maybe the extreme high or low temperature will cause this token?

**Reply by naailkhan28** [reply]
That's very strange, I've always found sequences with that low a temperature setting to have pretty good quality with no odd tokens. 

What kind of proteins are you designing? Maybe the model doesn't generalize very well to your structures and so assigns a higher probability to these odd tokens. 

Also how many sequences are you designing at a time? If you're making loads of sequences at a time it may be difficult to avoid one or two of these. 

I'm also not sure if <EOS> in particular might be a token that comes up in designed sequences a lot - maybe one of the team can shed some light on this.

**Reply by sirius777coder** [reply]
I design 100 sequences for one de novo designed protein backbone. And sometimes I may meet this problem when I design 1000 sequences or so.

**Reply by sirius777coder** [reply]
I design 100 sequences for one de novo designed protein backbone. And sometimes I may meet this problem when I design 1000 sequences or so.

**Comment by tomsercu** [comment, marked as answer]
Interesting observation.
Because it's an Autoregressive decoder there's no explicit supervision on when to emit EOS. However since the length is the same as the encoder input (backbone) length I would have expected the model to trivially learn this.. Apparently not exactly.
In neural machine translation there are some standard hacks going around to suppress sampling EOS too soon or too late (ie generate a sequence of appropriate length). In this case it'd be very reasonable to suppress EOS in the `sample()` function of `ems.inverse_folding.gvp_transformer.GVPTransformerModel`

**Reply by sirius777coder** [reply]
Thank you! I am curious about whether we can't really avoid the eos token? Compared with other models like ProteinSolver, ESM-IF1 is a really good program but I will still find 15 sequences when I design 15,000 sequences.
